// craiyon/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CraiyonModel = exports.CraiyonImage = exports.CraiyonOutput = exports.ClientV3 = exports.ClientV2 = exports.ClientV1 = exports.Client = void 0;
const clientv3_1 = __importStar(require("./api/clientv3"));
exports.Client = clientv3_1.default;
exports.ClientV3 = clientv3_1.default;
Object.defineProperty(exports, "CraiyonModel", { enumerable: true, get: function () { return clientv3_1.CraiyonModel; } });
const clientv2_1 = __importDefault(require("./api/clientv2"));
exports.ClientV2 = clientv2_1.default;
const clientv1_1 = __importDefault(require("./api/clientv1"));
exports.ClientV1 = clientv1_1.default;
const CraiyonOutput_1 = __importDefault(require("./models/CraiyonOutput"));
exports.CraiyonOutput = CraiyonOutput_1.default;
const CraiyonImage_1 = __importDefault(require("./models/CraiyonImage"));
exports.CraiyonImage = CraiyonImage_1.default;
