// craiyon/lib/models/CraiyonOutput.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const CraiyonImage_1 = __importDefault(require("./CraiyonImage"));
/**
 * @api public
 * @class CraiyonOutput
 * @classdesc Represents a Craiyon response.
 */
class CraiyonOutput {
    /**
     * @constructor
     * @param {CraiyonImage[]} images An array of image objects.
     * @param {string} version The version of the Craiyon model used to generate
     *                         the images.
     */
    constructor(images, version) {
        this._images = images;
        this._version = version;
    }
    /**
     * A factory function that will create a CraiyonOutput instance from the given
     * JSON response.
     *
     * @param {object} json The JSON response to parse.
     * @returns {CraiyonOutput} The parsed CraiyonOutput instance.
     */
    static fromJSON(json) {
        const images = json.images.map((image) => new CraiyonImage_1.default(image));
        return new CraiyonOutput(images, json.version);
    }
    /**
     * Will return the base64 representation of the images in the output.
     *
     * @returns {string[]} An array of base64-encoded images.
     */
    asBase64() {
        return this._images.map((image) => image.asBase64());
    }
    /**
     * Will return an array of buffers containing the images in the output.
     *
     * @returns {Buffer[]} An array of buffers containing the image data.
     */
    asBuffers() {
        return this._images.map((image) => image.asBuffer());
    }
    /**
     * Getter for the array of Craiyon image objects.
     *
     * @returns {CraiyonImage[]} The array of Craiyon image objects.
     */
    get images() {
        return this._images;
    }
    /**
     * Getter for the version of the Craiyon model used to generate the images.
     *
     * @returns {string} The version of the Craiyon model.
     */
    get version() {
        return this._version;
    }
}
exports.default = CraiyonOutput;
